File: 2022/d03.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <bitset>
#include <stdexcept>

// considered bad practice
using namespace std;

void mark_items(bitset<53> &items, const string &curline) {
    for (const char &c : curline) {
        int prio = c <= static_cast<int>('Z') ?
            c - static_cast<int>('A') + 26 :
            c - static_cast<int>('a');
        items.set(prio);
    }
}

// NOTE: msvc only shows unhandled exceptions in the debugger
// otherwise the program just crashes
int main() {
    ifstream infile("d03.in");

    string curline;
    int prio_sum = 0;
    while (getline(infile, curline)) {
        if (curline.size() == 0) continue;

        int half_idx = curline.size() / 2;
        string compartment1 = curline.substr(0, half_idx);
        string compartment2 = curline.substr(half_idx, half_idx);
        // cout << compartment1 << endl << compartment2 << endl;
        // a-z prio: 1--26
        // A-Z prio: 27--52
        bitset<53> items1;
        bitset<53> items2;
        mark_items(items1, compartment1);
        mark_items(items2, compartment2);
        auto result = items1 & items2;
        // iterate over result bitset adding the prio (idx in bitset + 1) if item was in both
        // compartments (is true after &)
        for (size_t i = 0; i < result.size(); ++i) {
            if (result[i]) {
                prio_sum += i + 1;
            }
        }
    }

    // part2
    // return to beginning of file and clear eof flags etc.
    infile.clear();
    infile.seekg(0);

    // start at 1 so we don't hit 0%3==0 instantly
    int idx = 1;
    int common_prio_sum = 0;
    bitset<53> rucksacks[3];
    while (getline(infile, curline)) {
        if (curline.size() == 0) continue;

        // make sure this is a reference or use a pointer!
        auto &items = rucksacks[idx % 3];
        mark_items(items, curline);

        if ((idx % 3) == 0) {
            auto result = (rucksacks[0] & rucksacks[1]) & rucksacks[2];
            for (size_t i = 0; i < result.size(); ++i) {
                if (result[i]) {
                    common_prio_sum += i + 1;
                }
            }

            // reset the 3 bitsets
            rucksacks[0].reset();
            rucksacks[1].reset();
            rucksacks[2].reset();
        }
        idx += 1;
    }

    cout << "Part1: Duplicate items prio sum is " << prio_sum << endl;
    cout << "Part2: Triple items prio sum is " << common_prio_sum << endl;
}

File: 2022/d03.go
package main

import (
    "fmt"
    "bufio"
    "os"
    "strings"
)

func mark_items(items []bool, str string) {
    for _, c := range str {
        var idx int
        if c <= 'Z' {
            idx = int(c) - 'A' + 26
        } else {
            idx = int(c) - 'a'
        }

        // fmt.Println("C", c, "->", string(c), "I", idx)
        items[idx] = true
    }
}

func main() {
    b, err := os.Open("d03.in")
    if err != nil {
        fmt.Println(err)
    }

    scanner := bufio.NewScanner(b)
    point_sum := 0;
    var lines []string
    // iterate over lines
    for scanner.Scan() {
        // get line and trim whitespace
        ln := strings.TrimSpace(scanner.Text())
        lines = append(lines, ln)
        if len(ln) == 0 {
            continue;
        }

        half := len(ln) / 2;
        compartment1 := ln[:half];
        compartment2 := ln[half:];
        // initialized to 0
        var items1 [53]bool
        var items2 [53]bool
        mark_items(items1[:], compartment1)
        mark_items(items2[:], compartment2)

        for i := range items1 {
            in_both := items1[i] && items2[i]
            if in_both {
                point_sum += i + 1
            }
        }
    }
    if err := scanner.Err(); err != nil {
        fmt.Println("Error reading from file input:", err)
    }

    var rucksacks [3][53]bool
    // start at 1 due to %3
    idx := 1
    common_point_sum := 0
    for _, ln := range lines {
        items := &rucksacks[idx % 3]
        mark_items(items[:], ln)

        if idx % 3 == 0 {
            for i := range rucksacks[0] {
                in_all := rucksacks[0][i] && rucksacks[1][i] && rucksacks[2][i]
                if in_all {
                    common_point_sum += i + 1
                }
            }

            // reset
            rucksacks = [3][53]bool{}
        }

        idx += 1
    }

    fmt.Println("Part1: Duplicate item prios are", point_sum)
    fmt.Println("Part2: Triple item prios are", common_point_sum)
}

